moveit_ros/moveit_servo/src/servo.cpp

```diff
@@ -33,74 +33,404 @@
+  servo_status_ = StatusCode::NO_WARNING;
+
+  // Update the parameters
+  updateParams();
+
+  // Get the robot state and joint model group info.
+  moveit::core::RobotStatePtr robot_state = planning_scene_monitor_->getStateMonitor()->getCurrentState();
+  const moveit::core::JointModelGroup* joint_model_group =
+      robot_state->getJointModelGroup(servo_params_.move_group_name);
+
+  // Get necessary information about joints
+  const std::vector<std::string> joint_names = joint_model_group->getActiveJointModelNames();
+  moveit::core::JointBoundsVector joint_bounds = joint_model_group->getActiveJointModelsBounds();
+  const int num_joints = joint_names.size();
+
+  // State variables
+  KinematicState current_state(num_joints), target_state(num_joints);
+  target_state.joint_names = joint_names;
+
+  // Copy current kinematic data from RobotState.
+  robot_state->copyJointGroupPositions(joint_model_group, current_state.positions);
+  robot_state->copyJointGroupVelocities(joint_model_group, current_state.velocities);
+
+  // Create Eigen maps for cleaner operations.
+  Eigen::Map<Eigen::VectorXd> current_joint_positions(current_state.positions.data(), num_joints);
+  Eigen::Map<Eigen::VectorXd> target_joint_positions(target_state.positions.data(), num_joints);
+  Eigen::Map<Eigen::VectorXd> current_joint_velocities(current_state.velocities.data(), num_joints);
+  Eigen::Map<Eigen::VectorXd> target_joint_velocities(target_state.velocities.data(), num_joints);
+
+  // Compute the change in joint position due to the incoming command
+  Eigen::VectorXd joint_position_delta = jointDeltaFromCommand(command, robot_state);
+
+  if (collision_velocity_scale_ > 0 && collision_velocity_scale_ < 1)
+  {
+    servo_status_ = StatusCode::DECELERATE_FOR_COLLISION;
+  }
+  else if (collision_velocity_scale_ == 0)
+  {
+    servo_status_ = StatusCode::HALT_FOR_COLLISION;
+  }
+
+  // Continue rest of the computations only if the command is valid
+  // The computations can be skipped also in case we are halting.
+  if (servo_status_ != StatusCode::INVALID || servo_status_ != StatusCode::HALT_FOR_COLLISION)
+  {
+    // Apply collision scaling to the joint position delta
+    joint_position_delta *= collision_velocity_scale_;
+
+    // Compute the next joint positions based on the joint position deltas
+    target_joint_positions = current_joint_positions + joint_position_delta;
+
+    // TODO : apply filtering to the velocity instead of position
+    // Apply smoothing to the positions if a smoother was provided.
+    // Update filter state and apply filtering in position domain
+    if (smoother_)
+    {
+      smoother_->reset(current_state.positions);
+      smoother_->doSmoothing(target_state.positions);
+    }
+
+    // Compute velocities based on smoothed joint positions
+    target_joint_velocities = (target_joint_positions - current_joint_positions) / servo_params_.publish_period;
+
+    // TODO : print warning if scaling applied for joint limit.
```

Comment:
Is this TODO for this PR? Seems like a quick one, so probably?

Reply:
Done